util/cpuprofile/pprof_api.go

```diff
@@ -0,0 +1,196 @@
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", `attachment; filename="profile"`)
+
+	pc := NewCollector()
+	err = pc.StartCPUProfile(w)
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+		return
+	}
+	time.Sleep(time.Second * time.Duration(sec))
+	err = pc.StopCPUProfile()
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+	}
+}
+
+const (
+	labelSQLDigest  = "sql_digest"
+	labelPlanDigest = "plan_digest"
+)
+
+// Collector is a cpu profile collector, it collect cpu profile data from globalCPUProfiler.
+type Collector struct {
+	ctx       context.Context
+	cancel    context.CancelFunc
+	started   atomic.Bool
+	firstRead chan struct{}
+	wg        sync.WaitGroup
+
+	dataCh   ProfileConsumer
+	profiles []*ProfileData
+	writer   io.Writer
+}
+
+// NewCollector returns a new NewCollector.
+func NewCollector() *Collector {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Collector{
+		ctx:       ctx,
+		cancel:    cancel,
+		firstRead: make(chan struct{}),
+		dataCh:    make(ProfileConsumer, 1),
+	}
+}
+
+// StartCPUProfile is a substitute for the `pprof.StartCPUProfile` function.
+// You should use this function instead of `pprof.StartCPUProfile`.
+// Otherwise you may fail, or affect the TopSQL feature and pprof profile HTTP API .
+func (pc *Collector) StartCPUProfile(w io.Writer) error {
+	if !pc.started.CAS(false, true) {
+		return errors.New("Collector already started")
+	}
+	pc.writer = w
+	pc.wg.Add(1)
+	go goutil.WithRecovery(pc.readProfileData, nil)
+	return nil
+}
+
+// StopCPUProfile is a substitute for the `pprof.StopCPUProfile` function.
+func (pc *Collector) StopCPUProfile() error {
+	if !pc.started.Load() {
+		return nil
+	}
+
+	// wait for reading least 1 profile data.
+	<-pc.firstRead
+	pc.cancel()
+	pc.wg.Wait()
+
+	data, err := pc.buildProfileData()
+	if err != nil {
+		return err
+	}
+	return data.Write(pc.writer)
+}
+
+// WaitProfilingFinish waits for collecting `seconds` profile data finished.
+func (pc *Collector) readProfileData() {
+	// register cpu profile consumer.
+	Register(pc.dataCh)
+	defer func() {
+		Unregister(pc.dataCh)
+		pc.wg.Done()
+	}()
+
+	pc.profiles = []*ProfileData{}
+	for {
+		select {
+		case <-pc.ctx.Done():
+			return
+		case data := <-pc.dataCh:
+			pc.profiles = append(pc.profiles, data)
+			if len(pc.profiles) == 1 {
+				close(pc.firstRead)
+			}
+		}
+	}
+}
+
+func (pc *Collector) buildProfileData() (*profile.Profile, error) {
+	ds := make([]*profile.Profile, 0, len(pc.profiles))
+	for _, data := range pc.profiles {
+		if data.Error != nil {
+			return nil, data.Error
+		}
+		pd, err := profile.ParseData(data.Data.Bytes())
+		if err != nil {
+			return nil, err
+		}
+		ds = append(ds, pd)
+	}
+
+	profileData, err := profile.Merge(ds)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Maybe you can merge them one by one and there is no need to build extra array to store them?

Reply:
great, done.